src/common/utils.cc

```diff
@@ -0,0 +1,61 @@
+// Copyright (C) 2019-2023 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#include "knowhere/utils.h"
+
+#include <cmath>
+#include <cstdint>
+
+#include "knowhere/comp/time_recorder.h"
+#include "knowhere/log.h"
+
+namespace knowhere {
+
+const float floatDiff = 0.00001;
+
+void
+normalize(float* data, int32_t dim) {
+    double vecLen = 0.0;
+    for (int32_t j = 0; j < dim; j++) {
+        vecLen += data[j] * data[j];
```

Comment:
float overflow risky, maybe we can do `a += (double)x * (double)c`

Reply:
updated